#2ndGS_Pipelines                                                          
###Perl 5 and R source codes for next generation (second generation) sequencing data analysis by integrating lots of open-source softwares.
                                                             

+ `CISDA`: ChIP-Seq Data Analyzer (also contains MNase-Seq, DNase-Seq, ChIP-exo and all ChIPseq-like data.)                                              
                                                                  
+ `MESDA`: MethylC-Seq Data Analyzer (also contains Bisulfite Sequencing and Single-cell BS-seq data.)                         
                       
+ `RASDA`: RNA-Seq Data Analyzer  
                                       
+ `HISDA`: Hi-C  Data Analyzer            
                                                               
                                                               
  All the scripts have been fully tested and passed my quality inspection in Ubuntu Linux.                  
                                               
#[Download](https://github.com/CTLife/2ndGS_Pipelines/releases)                   
---------------------------------------------------------------------------------------------                                                                     
### To use all the above 4 pipelines, the recent versions of these softwares must be available in your Linux OS.          
####Common Tools:                                        
1. [Perl 5](https://www.perl.org/) , [R](https://www.r-project.org/) , [NCBI SRA Toolkit](http://www.ncbi.nlm.nih.gov/Traces/sra/)   
4. bzip2, gunzip, tar, unrar, xz and unzip.  (You can install them by using "sudo apt-get isntall" in Ubuntu.)      
5. [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)     #QC of fastq files (5-24)                    
6. [NGS QC Toolkit](http://www.nipgr.res.in/ngsqctoolkit.html) 
7. [FASTX-Toolkit](http://hannonlab.cshl.edu/fastx_toolkit/)      
8. [Ea-utils](https://launchpad.net/ubuntu/+source/ea-utils)         
9. [kPAL](https://github.com/LUMC/kPAL)
10. [FaQCs](https://github.com/chienchi/FaQCs)    
11. [prinseq](http://prinseq.sourceforge.net/)      
12. [SolexaQA](http://solexaqa.sourceforge.net/)     
13. [fqtools](https://github.com/alastair-droop/fqtools)        
14. [fastq tools](http://homes.cs.washington.edu/~dcjones/fastq-tools/)                    
15. [HTSeq](https://github.com/bowhan/HTSeq-count)          
16. [HTQC](https://sourceforge.net/projects/htqc/files/)    
17. [fastqp](https://github.com/mdshw5/fastqp)    
18. [Rqc](http://bioconductor.org/packages/release/bioc/html/Rqc.html)       
19. [ShortRead](https://bioconductor.org/packages/release/bioc/html/ShortRead.html)     
20. [seqTools](https://www.bioconductor.org/packages/release/bioc/html/seqTools.html)              
21. [MultiQC](http://multiqc.info/)     
22. [raspberry](https://github.com/CEG-ICRISAT/NGS-QCbox)       
23. [ngsutils](http://ngsutils.org/)                   
24. [QC3](https://github.com/slzhao/QC3)        #QC of fastq files (5-24)      
25. [fastools](https://pypi.python.org/pypi/fastools)  (convert fastq to fasta)  
26. [TOGGLE](https://github.com/SouthGreenPlatform/TOGGLE)         
25. [Skewer](https://sourceforge.net/projects/skewer/files/)                   
26. [AdapterRemoval](https://github.com/MikkelSchubert/adapterremoval)                         
27. [Cutadapt](https://github.com/marcelm/cutadapt)                                   
28. [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)                  


####Only for CISDA: 
1. [Stampy](http://www.well.ox.ac.uk/bioinformatics/Software/Stampy-latest.tgz)   
14. MACS1.4, http://liulab.dfci.harvard.edu/MACS/         
15. MACS2, https://github.com/taoliu/MACS/                
16. HOMER, http://homer.salk.edu/homer/                
17. DANPOS, http://code.google.com/p/danpos/      
18. Peak Caller Q: http://charite.github.io/Q/                           
19. bamUtil:  http://ngsutils.org/modules/bamutils/        
20. HISAT2: http://ccb.jhu.edu/software/hisat2
21. R: https://www.r-project.org/
22. Picard: http://broadinstitute.github.io/picard/                                
23. Qualimap 2: http://qualimap.bioinfo.cipf.es/
24. STAR: https://github.com/alexdobin/STAR/releases                                                                      
25. SAMStat: http://samstat.sourceforge.net/                                                          
26. BamTools: https://github.com/pezmaster31/bamtools/wiki/Building-and-installing  
27. 6. Subread, http://subread.sourceforge.net/               
7. BWA, http://bio-bwa.sourceforge.net/          
8. Bowtie, http://bowtie-bio.sourceforge.net/index.shtml       
9. Bowtie2, http://bowtie-bio.sourceforge.net/bowtie2/index.shtml              
10. SAMtools, http://www.htslib.org/            
11. BEDtools, https://github.com/arq5x/bedtools2/releases                                
12. BEDOPS, http://bedops.readthedocs.org/en/latest/                
13. Deeptools, http://deeptools.github.io/ 
                                  
####Only for MESDA:                            
                               
####Only for RASDA:                         
                            
####Only for HISDA:                                                  
                                                                                           
                                                                                                        
---------------------------------------------------------------------------------
###License: The GNU General Public License v3.0                    
                                                                         
