# 2ndGS_Pipelines                          
###Perl source codes for next generation (second generation) sequencing data analysis by integrating lots of open-source softwares.
                                                             
`CISDA`: ChIP-Seq Data Analyzer (also contains MNase-seq, DNase-seq, ChIP-exo and all ChIPseq-like data.)                                              
                                                                  
`MESDA`: MethylC-Seq Data Analyzer    
                       
`RASDA`: RNA-Seq Data Analyzer  
                                       
`HISDA`: Hi-C  Data Analyzer            
                                                               
                                                               
  All the scripts have been fully tested and passed my quality inspection in Ubuntu Linux.                  
                                               
####Download: https://github.com/CTLife/2ndGS_Pipelines/releases                   
---------------------------------------------------------------------------------------------                                                                     
### To use all the above 4 pipelines, the recent versions of these softwares must be available in your Linux OS: 
0. Perl, https://www.perl.org/                 
1. NCBI SRA Toolkit, http://www.ncbi.nlm.nih.gov/Traces/sra/                                
2. FastQC, http://www.bioinformatics.babraham.ac.uk/projects/fastqc/
3. Trimmomatic, http://www.usadellab.org/cms/?page=trimmomatic                   
4. NGS QC Toolkit, http://www.nipgr.res.in/ngsqctoolkit.html              
5. FASTX-Toolkit, http://hannonlab.cshl.edu/fastx_toolkit/                  
6. Subread, http://subread.sourceforge.net/               
7. BWA, http://bio-bwa.sourceforge.net/          
8. Bowtie, http://bowtie-bio.sourceforge.net/index.shtml       
9. Bowtie2, http://bowtie-bio.sourceforge.net/bowtie2/index.shtml              
10. SAMtools, http://www.htslib.org/            
11. BEDtools, https://github.com/arq5x/bedtools2/releases                                
12. BEDOPS, http://bedops.readthedocs.org/en/latest/                
13. Deeptools, http://deeptools.github.io/                     
14. MACS1.4, http://liulab.dfci.harvard.edu/MACS/         
15. MACS2, https://github.com/taoliu/MACS/                
16. HOMER, http://homer.salk.edu/homer/                
17. DANPOS, http://code.google.com/p/danpos/      
18. Peak Caller Q: http://charite.github.io/Q/                           
19. bamUtil:  http://ngsutils.org/modules/bamutils/        
20. HISAT2: http://ccb.jhu.edu/software/hisat2
21. R: https://www.r-project.org/
22. Picard: http://broadinstitute.github.io/picard/                                
23. Qualimap 2: http://qualimap.bioinfo.cipf.es/
24. STAR: https://github.com/alexdobin/STAR/releases                                                                      
25. SAMStat: http://samstat.sourceforge.net/                                                          
26. BamTools: https://github.com/pezmaster31/bamtools/wiki/Building-and-installing                        

                                                                                           
                                                                                                        
---------------------------------------------------------------------------------
###License: The GNU General Public License v3.0                    
                                                                         
